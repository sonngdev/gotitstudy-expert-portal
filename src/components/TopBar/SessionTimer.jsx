import React, { useState, useEffect } from 'react';
import { formatCountdownTimer } from '../../helpers';

export function SessionTimer() {
  const [ timeLeft, setTimeLeft ] = useState(300000);
  const timesUp = timeLeft < 1000;

  useEffect(() => {
    if (timesUp) window.location.reload();

    const interval = setInterval(() => setTimeLeft(timeLeft - 1000), 1000);
    return () => clearInterval(interval);
  });

  return (
    <div className={`session-timer ${timesUp ? 'u-backgroundNegative' : 'u-backgroundPositive'} u-roundedLarge u-paddingTiny u-paddingHorizontalExtraSmall u-text100 u-textUppercase u-textWhite`}>
      Discussion session active
      <div className="stat u-text600 u-textCenter">{formatCountdownTimer(timeLeft)}</div>
    </div>
  );
}
