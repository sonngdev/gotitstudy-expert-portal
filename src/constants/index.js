export const DEFAULT_TIMEZONE = 'America/Chicago';
/**
 * `z` is deprecated in moment, but not in moment-timezone.
 *
 * We need to use `z` in our format, but we cannot test the output by trying
 * to parse it back (using moment) to see if it is valid. This is because to
 * moment (at least in newer versions) it is not.
 *
 */
export const LONG_TIME_FORMAT = 'MMM D, YYYY HH:mm:ss z';
/**
 * A session is 5 minutes (300,000 miliseconds) long.
 */
export const SESSION_DURATION = 300000;
