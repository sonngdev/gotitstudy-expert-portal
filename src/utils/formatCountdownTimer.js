export function formatCountdownTimer(milisecond) {
  const minute = Math.floor(milisecond / 1000 / 60);
  const second = Math.floor(milisecond / 1000) - 60 * minute;

  return [minute, second].map((n) => String(n).padStart(2, '0')).join(':');
}
