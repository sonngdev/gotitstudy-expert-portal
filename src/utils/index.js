export { getCurrentTime, formatCountdownTimer, formatCurrentTime } from './time';
export { useChatMessagesList, useChatMessagesAdd } from './useChatMessages';
export { useScrollToBottom } from './useScrollToBottom';
export { useTinyMceEditor } from './useTinyMceEditor';
