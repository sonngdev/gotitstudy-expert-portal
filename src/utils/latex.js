/* eslint-disable prefer-arrow-callback */

tinymce.PluginManager.add('latex', function addLatexPlugin(editor) {
  /**
  |--------------------------------------------------
  | VARIABLES
  |--------------------------------------------------
  */
  const latexClass = 'Latexformula';

  const latexPreviewId = `${latexClass}-preview`;


  /**
  |--------------------------------------------------
  | HELPERS
  |--------------------------------------------------
  */
  const isLatexFormula = (node) => [...node.classList].includes(latexClass);

  const getExpression = (dialog) => dialog.getData().expression.trim();

  const getSelectedExpression = () => {
    const { selection } = editor;
    const node = selection && selection.getNode();
    return node && isLatexFormula(node) ? node.alt : '';
  };

  const getFormulaSrc = (expression) => (
    /**
     * LaTeX Math as SVG image:
     * https://math.now.sh/home
     */
    `https://math.now.sh?from=${encodeURIComponent(expression)}`
  );

  const setPreview = (content) => {
    document.getElementById(latexPreviewId).innerHTML = content;
  };


  /**
  |--------------------------------------------------
  | DIALOG
  |--------------------------------------------------
  */
  const openDialog = () => editor.windowManager.open({

    title: 'LaTeX expression',

    body: {
      type: 'panel',
      items: [
        { type: 'textarea', name: 'expression', label: 'Expression' },
        { type: 'button', name: 'preview', text: 'Preview' },
        { type: 'htmlpanel', html: `<div id="${latexPreviewId}"></div>` },
      ],
    },

    buttons: [
      { type: 'cancel', text: 'Cancel' },
      { type: 'submit', text: 'Insert', primary: true },
    ],

    initialData: {
      expression: getSelectedExpression(),
    },

    onAction(dialog, details) {
      if (details.name !== 'preview') return;

      const expression = getExpression(dialog);
      if (!expression) return;

      const src = getFormulaSrc(expression);
      const content = `<img src="${src}" alt="${expression}" role="math" />`;
      setPreview(content);
    },

    onSubmit(dialog) {
      const expression = getExpression(dialog);
      if (!expression) return;

      const src = getFormulaSrc(expression);
      const content = `<img class="${latexClass}" src="${src}" alt="${expression}" role="math"/>`;
      editor.insertContent(content);

      dialog.close();
    },
  });


  /**
  |--------------------------------------------------
  | REGISTER
  |--------------------------------------------------
  */
  editor.ui.registry.addButton('latex', {
    text: 'LaTeX',
    onAction: openDialog,
  });

  editor.ui.registry.addMenuItem('latex', {
    text: 'LaTeX',
    onAction: openDialog,
  });

  editor.on('DblClick', ({ target }) => {
    if (isLatexFormula(target)) openDialog();
  });

  return {
    getMetadata: () => ({ name: 'LaTeX plugin for tinymce 5' }),
  };
});
