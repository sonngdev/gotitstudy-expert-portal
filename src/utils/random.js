/**
 * For trivial randomness, this solution is sufficient
 * https://stackoverflow.com/questions/1349404/generate-random-string-characters-in-javascript
 */
export function weakRandomString(len) {
  if (len >= 10) throw new Error('String length must be smaller than 10');
  return Math.random().toString(36).substring(2, 2 + len);
}
