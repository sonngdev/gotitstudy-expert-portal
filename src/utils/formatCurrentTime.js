import moment from 'moment-timezone';
import { DEFAULT_TIMEZONE, CURRENT_TIME_FORMAT } from '../constants';

export function formatCurrentTime() {
  return moment().tz(DEFAULT_TIMEZONE).format(CURRENT_TIME_FORMAT);
}
