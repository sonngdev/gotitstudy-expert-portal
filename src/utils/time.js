import moment from 'moment-timezone';
import { DEFAULT_TIMEZONE, LONG_TIME_FORMAT } from '../constants';

export function getCurrentTime() {
  return moment().tz(DEFAULT_TIMEZONE);
}

export function formatLongTime(momentObj) {
  return momentObj.format(LONG_TIME_FORMAT);
}

export function formatCountdownTimer(milisecond) {
  const absMilisecond = Math.abs(milisecond);

  const minute = Math.floor(absMilisecond / 1000 / 60);
  const second = Math.floor(absMilisecond / 1000) - 60 * minute;

  const sign = milisecond < 0 ? '-' : '';
  const min = String(minute);
  const sec = String(second).padStart(2, '0');

  return `${sign}${min}:${sec}`;
}
