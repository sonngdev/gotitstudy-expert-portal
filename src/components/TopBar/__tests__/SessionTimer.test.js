import React, * as fromReact from 'react';
import { shallow } from 'enzyme';
import { SessionTimer } from '../SessionTimer';
import * as fromUtils from '../../../constants';
import { formatCountdownTimer } from '../../../utils';

describe('<SessionTimer />', () => {
  let wrapper;
  /**
   * Mocking SESSION_DURATION because jest's `advanceTimersByTime`
   * does not seem to work well with large numbers.
   */
  const SESSION_DURATION = 10000;

  beforeEach(() => {
    jest.clearAllTimers();
    jest.useFakeTimers();
    /**
     * Another way to mock only part of a module.
     * The first way was documented in CurrentTime.test.js
     */
    fromReact.useEffect = jest.fn((fn) => fn());
    fromUtils.SESSION_DURATION = SESSION_DURATION;

    wrapper = shallow(<SessionTimer />);
  });

  it('renders', () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it('starts off at the default session duration', () => {
    expect(wrapper.find('.stat').text())
      .toBe(formatCountdownTimer(SESSION_DURATION));
  });

  it('counts down as time passes', () => {
    jest.advanceTimersByTime(2000);
    expect(wrapper.find('.stat').text())
      .toBe(formatCountdownTimer(SESSION_DURATION - 2000));

    jest.advanceTimersByTime(4000);
    expect(wrapper.find('.stat').text())
      .toBe(formatCountdownTimer(SESSION_DURATION - 6000));
  });

  it('reloads the page when time is up', () => {
    const { location } = window;
    delete window.location;
    window.location = { reload: jest.fn() };

    jest.advanceTimersByTime(SESSION_DURATION + 1000);
    expect(window.location.reload).toHaveBeenCalled();

    window.location = location;
  });

  it('has green background at the beginning', () => {
    expect(wrapper.hasClass('u-backgroundPositive')).toBe(true);
  });

  it('has red background at the end', () => {
    const { location } = window;
    delete window.location;
    window.location = { reload: jest.fn() };

    jest.advanceTimersByTime(SESSION_DURATION + 1000);
    expect(wrapper.hasClass('u-backgroundNegative')).toBe(true);

    window.location = location;
  });
});
