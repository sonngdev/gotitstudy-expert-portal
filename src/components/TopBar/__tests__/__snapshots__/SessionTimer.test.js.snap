// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SessionTimer /> renders 1`] = `
"<div className=\\"session-timer u-backgroundPositive u-roundedLarge u-paddingTiny u-paddingHorizontalExtraSmall u-text100 u-textUppercase u-textWhite\\">
  Discussion session active
  <div className=\\"stat u-text600 u-textCenter\\">
    0:10
  </div>
</div>"
`;
